source/Halibut.Tests/Transport/Streams/ReadIntoMemoryBufferStreamFixture.cs

```diff
@@ -0,0 +1,157 @@
+            using var memoryStream = CreateMemoryStreamPopulatedWith(bytesToWrite);
+            using var sut = new ReadIntoMemoryBufferStream(memoryStream, bytesToWrite.Length - 1, OnDispose.LeaveInputStreamOpen);
+
+            await sut.BufferIntoMemoryFromSourceStreamUntilLimitReached(CancellationToken);
+
+            // Act
+            var readBuffer = new byte[bytesToWrite.Length];
+            // First round read till we hit the memory buffer
+            var bytesRead = await ReadFromStream(streamMethod, sut, readBuffer, 0, bytesToWrite.Length);
+            bytesRead.Should().Be(bytesToWrite.Length - 1);
+
+            // Next round finishes the data from the source
+            bytesRead = await ReadFromStream(streamMethod, sut, readBuffer, bytesToWrite.Length - 1, 1);
+            bytesRead.Should().Be(1);
+
+            // Assert
+            readBuffer.Should().BeEquivalentTo(bytesToWrite);
+            sut.BytesReadIntoMemory.Should().Be(bytesToWrite.Length - 1);
+        }
+        
+        [Test]
+        public async Task PartiallyReadsFromMemoryBuffer_IfBufferingWasApplied_ReadingOneByteAtATime([Values] StreamMethod streamMethod)
+        {
+            // Arrange
+            var bytesToWrite = Encoding.ASCII.GetBytes("Some bytes for testing");
+            var readIntoMemoryLimitBytes = bytesToWrite.Length / 2;
+            using var memoryStream = CreateMemoryStreamPopulatedWith(bytesToWrite);
+            using var sut = new ReadIntoMemoryBufferStream(memoryStream, readIntoMemoryLimitBytes, OnDispose.LeaveInputStreamOpen);
+
+            await sut.BufferIntoMemoryFromSourceStreamUntilLimitReached(CancellationToken);
+
+            // Act
+            var readBuffer = new byte[bytesToWrite.Length];
+            for (int i = 0; i < bytesToWrite.Length; i++)
+            {
+                var bytesRead = await ReadFromStream(streamMethod, sut, readBuffer, i, 1);
+                bytesRead.Should().Be(1);
+            }
+
+            // Assert
+            readBuffer.Should().BeEquivalentTo(bytesToWrite);
+            sut.BytesReadIntoMemory.Should().Be(readIntoMemoryLimitBytes);
+        }
+
+        static MemoryStream CreateMemoryStreamPopulatedWith(byte[] bytesToWrite)
+        {
+            var memoryStream = new MemoryStream();
+            
+            memoryStream.Write(bytesToWrite, 0, bytesToWrite.Length);
+            memoryStream.Position = 0;
+
+            return memoryStream;
+        }
+
+        async Task<int> ReadFromStream(StreamMethod streamMethod, ReadIntoMemoryBufferStream sut, byte[] readBuffer, int offset, int count)
+        {
+            switch (streamMethod)
+            {
+                case StreamMethod.Async:
+                    return await sut.ReadAsync(readBuffer, offset, count, CancellationToken);
+                case StreamMethod.Sync:
+                    return sut.Read(readBuffer, offset, count);
+                case StreamMethod.BeginEnd:
```

Comment:
I don't exactly understand what is happening here, could a comment be added and/or the enum give more detail on what is happening.

Reply:
This is the legacy way .net used to do async.

I tested this as it was still a supported way to 'read' from a stream.

To help with understanding, I added a comment and renamed the enum to `LegacyAsync`